File: app/javascript/controllers/cart_controller.js
// app/javascript/controllers/cart_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["button", "delete"]


  add(event) {
    event.preventDefault()
    const engineId = this.buttonTarget.dataset.engineId

    fetch(`/cart/add/${engineId}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-CSRF-Token": document.querySelector("[name='csrf-token']").content
      },
      body: JSON.stringify({ engine_id: engineId })
    })
      .then(response => response.json())
      .then(data => {
        // update button text or cart badge
        this.buttonTarget.innerText = "Added"
        this.updateCartBadge(data.cart_count)
      })
  }

    remove(event) {
      event.preventDefault()

      const engineId = event.currentTarget.dataset.engineId // ✅ Must come first

      fetch(`/cart/remove/${engineId}`, {
        method: "DELETE",
        headers: {
          "X-CSRF-Token": document.querySelector("[name='csrf-token']").content
        }
      })
        .then(response => response.json())
        .then(data => {
          // Remove the item from DOM
          const itemEl = event.currentTarget.closest(".cart-item")
          if (itemEl) itemEl.remove()

          this.updateCartBadge(data.cart_count)

          if (data.cart_count === 0) {
            const panel = document.querySelector("#cart-dropdown")
            if (panel) panel.innerHTML = "<p>Your cart is empty.</p>"
          }
        })
    }


  updateCartBadge(count) {
    const badge = document.querySelector("[data-cart-badge-target='count']")
    if (badge) badge.textContent = count
  }
}

File: app/controllers/cart_controller.rb
class CartController < ApplicationController
  def add
    session[:cart] ||= []
    session[:cart] << params[:engine_id] unless session[:cart].include?(params[:engine_id])

    render json: {
        status: "Added",
        cart: session[:cart],
        cart_count: session[:cart].size
    }

  end

  def remove
    session[:cart] ||= []
    session[:cart].delete(params[:id].to_i)

    respond_to do |format|
      format.turbo_stream { render turbo_stream: turbo_stream.replace("cart-dropdown", partial: "shared/dropdown") }
      format.html { redirect_to root_path, notice: "Item removed" }
    end
  end

  def show
    @engines = Engine.where(id: session[:cart] || [])
  end
end
